rust/lance/src/dataset/builder.rs

```diff
@@ -105,15 +105,43 @@ impl DatasetBuilder {
     /// - [S3 options](https://docs.rs/object_store/latest/object_store/aws/enum.AmazonS3ConfigKey.html#variants)
     /// - [Google options](https://docs.rs/object_store/latest/object_store/gcp/enum.GoogleConfigKey.html#variants)
     pub fn with_storage_options(mut self, storage_options: HashMap<String, String>) -> Self {
-        self.storage_options = Some(storage_options);
+        self.options.storage_options = Some(storage_options);
         self
     }
 
     /// Set a single option used to initialize storage backend
+    /// For example, to set the region for S3, you can use:
+    ///
+    /// ```ignore
+    /// let builder = DatasetBuilder::from_uri("s3://bucket/path")
+    ///     .with_storage_option("region", "us-east-1");
+    /// ```
     pub fn with_storage_option(mut self, key: impl AsRef<str>, value: impl AsRef<str>) -> Self {
-        let mut storage_options = self.storage_options.unwrap_or_default();
+        let mut storage_options = self.options.storage_options.unwrap_or_default();
         storage_options.insert(key.as_ref().to_string(), value.as_ref().to_string());
-        self.storage_options = Some(storage_options);
+        self.options.storage_options = Some(storage_options);
+        self
+    }
+
+    /// Set options based on [ReadParams].
+    pub fn with_read_params(mut self, read_params: ReadParams) -> Self {
+        self = self
+            .with_index_cache_size(read_params.index_cache_size)
+            .with_metadata_cache_size(read_params.metadata_cache_size);
+
+        if let Some(options) = read_params.store_options {
+            self.options = options;
+        }
+        self
+    }
+
+    /// Re-use an existing session.
+    ///
+    /// The session holds caches for index and metadata.
+    ///
+    /// If this is set, then `with_index_cache_size` and `with_metadata_cache_size` are ignored.
+    pub fn with_session(mut self, session: Arc<Session>) -> Self {
+        self.session = Some(session);
         self
     }
```

Comment:
I went back and confirmed and I think this is safe to use across multiple datasets.